quinn-proto/src/connection/mod.rs

```diff
@@ -693,17 +693,15 @@ impl Connection {
                         let packet_len_unpadded = cmp::max(builder.min_size, buf.len())
                             - datagram_start
                             + builder.tag_len;
-                        if packet_len_unpadded + MAX_PADDING < segment_size {
+                        if packet_len_unpadded + MAX_PADDING >= segment_size {
```

Comment:
I expect this condition to be unsatisfiable. `segment_size` is the maximum length packet assembly is permitted to produce.

Reply:
I'm sorry that you may be confusing. If this condition is always unsatisfiable, then the original condition is always true, and it means that we always execute original `break` operation in its branch, which is incorrect.

In fact, `segment_size` is exactly the maximum length, so it is always larger than `packet_len_unpadded`. But the comparison add extra `MAX_PADDING` to judge whether we skip amount of padding. Therefore, their sum may be larger than `segment_size` and then the condition can be satisfiable is some case.

In original version, if the condition is satisfiable, we directly `break` from the `while` loop  and execute the packet logic for the last packet. In current version, we just skip the padding operation  `builder.pad_to(segment_size as u16);` but keep all the following operations. 

In short, the original version skip all the following operations, but the current version just skip  the padding operation  `builder.pad_to(segment_size as u16);`.

Reply:
Please note that the involved commit is d5fc528930c0761215fcad13a7797e9904ae4b6a.

Reply:
> If this condition is always unsatisfiable, then the original condition is always true

Oops, yes, you're correct. Nonetheless, the original behavior is intended. The difference achieved by breaking here is what terminates the GSO batch, which is the purpose of this condition. Removing the `break` allows the batch to continue with incorrect padding.